File: algorithms/0583.delete_operation_for_two_strings/delete_operation_for_two_strings.cpp
class Solution {
public:
    int minDistance(const string& word1, const string& word2) {
        const size_t m = word1.length(), n = word2.length();
        vector<int> dp(n + 1);
        vector<int> last(n + 1);
        for (size_t i = 1; i <= n; ++i) {
            dp[i] = i;
        }
        for (size_t i = 1; i <= m; ++i) {
            dp.swap(last);
            dp[0] = i;
            for (size_t j = 1; j <= n; ++j) {
                if (word1[i - 1] == word2[j - 1]) {
                    dp[j] = last[j - 1];
                } else {
                    dp[j] = min(last[j], dp[j - 1]) + 1;
                }
            }
        }
        return dp[n];
    }
};

File: algorithms/0583.delete_operation_for_two_strings/delete_operation_for_two_strings.rs
impl Solution {
    pub fn min_distance(word1: String, word2: String) -> i32 {
        let (word1, word2) = (word1.as_bytes(), word2.as_bytes()); 
        let (n, m) = (word1.len(), word2.len());
        let mut last = vec![0; m + 1];
        let mut dp = vec![0; m + 1];
        for i in 1..=n {
            dp.swap_with_slice(&mut last);
            for j in 1..=m {
                dp[j] = if word1[i - 1] == word2[j - 1] {
                    last[j - 1] + 1
                } else {
                    last[j].max(dp[j - 1])
                };
            }
        }
        (n + m) as i32 - 2 * dp[m]
    }
}